File: primes/quadratic_sieve.py
# quadratic_sieve.py
# Trevor Pottinger
# Fri Dec  6 22:57:30 PST 2019

from __future__ import division
from __future__ import print_function

import argparse
import math
import sys

if sys.version_info >= (3, 3):
    from typing import (Callable, Dict, List, Optional, Tuple, TypeVar)
    greater_than_one = int
    greater_than_zero = int
    matrix = List[List[int]]
    maybe_prime = int
    nonnegative = int
    prime = int
    K = TypeVar('K')
    V = TypeVar('V')


def gcd(a, b):
    # type: (nonnegative, nonnegative) -> nonnegative
    """Based on https://en.wikipedia.org/wiki/Euclidean_algorithm#Procedure"""
    assert a >= 0, 'type violation, expected a >= 0'
    assert b >= 0, 'type violation, expected b >= 0'
    if b == 0:
        return a
    if a < b:
        return gcd(b, a)
    return gcd(b, a % b)


def pollardsRho(n, g):
    # type: (nonnegative, Optional[Callable[[nonnegative, nonnegative], nonnegative]]) -> Optional[nonnegative]
    # https://en.wikipedia.org/wiki/Pollard%27s_rho_algorithm
    if g is None:
        g = lambda x, mod: (x ** 2 + 1) % mod
    x = 2
    y = 2
    d = 1
    while d == 1:
        x = g(x, n)
        y = g(g(y, n), n)
        d = gcd(abs(x - y), n)
    if d == n:
        return None
    return d


# Translated from https://rosettacode.org/wiki/Integer_roots#C
def intSqrt(n):
    # type: (nonnegative) -> nonnegative
    """Returns the square root of n, rounded down"""
    assert n >= 0, 'type violation, expected n >= 0'
    c = 1
    d = (1 + n) // 2
    e = (d + n // d) // 2
    while c != d and c != e:
        c = d
        d = e
        e = (e + n // e) // 2
    if d < e:
        return d
    return e


def isSquare(n):
    # type: (nonnegative) -> Tuple[nonnegative, bool]
    # Alternative: factor `n`, and check that all the prime powers are even.
    root = intSqrt(n)
    return (root, root * root == n)


def fermatsMethod(n):
    # type: (nonnegative) -> Dict[maybe_prime, nonnegative]
    """Implements Fermat's method for factoring numbers. The idea is to find
    two numbers, a, and b, such that a ** 2 - b ** 2 == n. This form can be
    represented as (a - b) * (a + b), which therefore are two factors of n."""
    assert n >= 0, 'type violation, expected n >= 0'
    s = int(math.ceil(math.sqrt(n)))
    while True:
        d = s ** 2 - n
        d_root, d_is_square = isSquare(d)
        if d_is_square:
            return {s - d_root: 1, s + d_root: 1}
        s += 1
    return {}


def slowPrimes(n):
    # type: (greater_than_one) -> List[prime]
    """Returns all primes less than or equal to n"""
    assert n > 1, 'type violation, expected n > 1'
    if n == 2:
        return [2]
    primes = [2]
    for i in range(3, n + 1, 2):
        is_prime = True
        for p in primes:
            if i % p == 0:
                is_prime = False
                break
        if is_prime:
            primes.append(i)
    return primes


def slowFactors(n, primes=None):
    # type: (greater_than_one, Optional[List[prime]]) -> Tuple[bool, Dict[prime, greater_than_zero]]
    assert n > 1, 'type violation, expected n > 1'
    if primes is None:
        # generate all primes up to n, instead of up to ceil(sqrt(n)) b/c
        # we don't have logic to treat the remainder as a prime
        primes = slowPrimes(n)
    ret = {}
    for p in primes:
        exponent = 0
        while n % p == 0:
            n //= p
            exponent += 1
        if exponent == 0:
            continue
        ret[p] = exponent
        if n == 1:
            return (True, ret)
    # Return false because we didn't finish factoring
    return (False, ret)


def vectorize(rows, default=None):
    # type: (List[Dict[K, V]], Optional[V]) -> List[List[V]]
    """Given a list of dicts, returns a list of vectors. Where each vector is
    of the same length, and each key in the input dicts corresponds to the same
    column in the output. If the input is from JSON, this is effectively
    "unsparsifying" the input."""
    key_to_col = {}  # type: Dict[K, nonnegative]
    num_keys = 0
    # first pass for key discovery
    for row in rows:
        for k in row:
            if k not in key_to_col:
                key_to_col[k] = num_keys
                num_keys += 1
    ret = []  # type: List[List[V]]
    # second pass for constructing vectors
    for row in rows:
        vector = [default for _ in range(num_keys)]
        for k in row:
            vector[key_to_col[k]] = row[k]
        ret.append(vector)
    return ret


def modularRowReduction(matrix, mod):
    # type: (matrix, Optional[greater_than_one]) -> matrix
    # Based on https://rosettacode.org/wiki/Reduced_row_echelon_form#Python
    assert mod is None or mod > 1, 'type violation, expected mod > 1'
    nrows = len(matrix)
    assert nrows > 0, 'matrix must have at least one row'
    ncols = len(matrix[0])
    assert ncols > 0, 'matrix must have at least one column'

    lead = 0
    for i in range(nrows):
        if lead >= ncols:
            break
        j = i

        # find the first row, j, where col `lead` is non-zero
        while matrix[j][lead] == 0:
            j += 1
            if j < nrows:
                continue
            # there are no non-zero values in column `lead`!
            j = i
            lead += 1
            if lead == ncols:
                return matrix

        # swap rows i and j
        matrix[i], matrix[j] = matrix[j], matrix[i]
        lead_value = matrix[i][lead]
        # divide the row by lead_value so the lead value becomes 1
        matrix[i] = [val // lead_value for val in matrix[i]]

        # cancel out column `lead` for each row j, s.t. j != i
        for j in range(nrows):
            if i == j:
                continue
            lead_value = matrix[j][lead]
            row = []
            for j_value, i_value in zip(matrix[i], matrix[j]):
                if mod is not None:
                    row.append((i_value - lead_value * j_value) % mod)
                else:
                    row.append(i_value - lead_value * j_value)
            matrix[j] = row
        lead += 1

    return matrix


def matstr(mat):
    # type: (matrix) -> str
    return  "[" + ",\n ".join(map(str, mat)) + "]"


def matExtend(mat_a, mat_b):
    # type: (matrix, matrix) -> matrix
    nrows = len(mat_a)
    assert nrows > 0, 'mat_a must have at least one row'
    assert nrows == len(mat_b), 'mat_b must have the same nrows as mat_a'
    ncols = len(mat_a[0]) + len(mat_b[0])
    assert ncols > 0, 'mat_a + mat_b must have at least one column'
    ret = []
    for i in range(nrows):
        ret.append(mat_a[i] + mat_b[i])
    return ret


def identity(nrows, ncols=None):
    # type: (greater_than_zero, Optional[greater_than_zero]) -> matrix
    assert nrows > 0, 'type violation, expected nrows > 0'
    assert ncols is None or ncols > 0, 'type violation, expected ncols > 0'
    if ncols is None:
        ncols = nrows
    ret = []
    for i in range(nrows):
        ret.append([0 for _ in range(i)])
        if i == ncols:
            continue
        ret[i].append(1)
        ret[i].extend([0 for _ in range(ncols - i - 1)])
    return ret


def findSquareProduct(primes, ints):
    # type: (Optional[List[prime]], List[nonnegative]) -> Tuple[List[List[nonnegative]], List[nonnegative]]
    """Given a list of distinct integers, finds products of a subset of them
    that is a square and returns those products."""
    if primes is None:
        primes = slowPrimes(max(ints))
    rows = []
    for i in ints:
        factored, factors = slowFactors(i, primes)
        if not factored:
            continue
        rows.append(factors)
    exponents = vectorize(rows, default=0)
    even_exponents = []
    for vector in exponents:
        even_exponents.append([e % 2 for e in vector])

    # We might use less than all `primes`, as well as less than all ints
    n_primes = len(even_exponents[0])
    n_ints = len(even_exponents)
    res = modularRowReduction(matExtend(even_exponents, identity(n_ints)), 2)
    zero_row = [0 for _ in range(n_primes)]

    indices = []
    solutions = []
    for i in range(n_ints):
        if res[i][:n_primes] != zero_row:
            continue
        solution = 1
        for j, bit in enumerate(res[i][n_primes:]):
            if bit == 0:
                continue
            solution *= ints[j]
        indices.append(res[i][n_primes:])
        solutions.append(solution)

    return (indices, solutions)


def isBSmooth(primes, i):
    # type: (List[prime], greater_than_one) -> bool
    for p in primes:
        while i % p == 0:
            i //= p
        if i == 1:
            return True
    return False


def bSmoothList(primes, n, count):
    # type: (List[prime], greater_than_one, greater_than_one) -> Tuple[List[greater_than_one], List[greater_than_one]]
    """This isn't a real sieve. It's just really simple. Given an int, n,
    and the number of smooth numbers to try getting, count, derive the list
    of smooth ints."""
    n_root = intSqrt(n)
    # This ensures we take the ceiling of the sqrt
    if n_root * n_root < n:
        n_root += 1
    ints = []
    squares = []
    for i in range(n_root, n_root + count + 1):
        # `i ** 2 - n` should be equivalent to `i ** 2 % n` because of the
        # range that i is in, and therefore what `i ** 2` is in.
        square_mod_n = i ** 2 - n
        if not isBSmooth(primes, square_mod_n):
            continue
        ints.append(i)
        squares.append(square_mod_n)
    return (ints, squares)


# From https://rosettacode.org/wiki/Tonelli-Shanks_algorithm#Python
def legendre(a, p):
    return pow(a, (p - 1) // 2, p)


# From https://rosettacode.org/wiki/Tonelli-Shanks_algorithm#Python
def tonelli(n, p):
    if not legendre(n, p) == 1:
        # previously this asserted "not a square (mod p)"
        return (False, None)
    q = p - 1
    s = 0
    while q % 2 == 0:
        q //= 2
        s += 1
    if s == 1:
        return (True, pow(n, (p + 1) // 4, p))
    z = 2
    for i in range(2, p):
        z = i
        if p - 1 == legendre(z, p):
            break
    c = pow(z, q, p)
    r = pow(n, (q + 1) // 2, p)
    t = pow(n, q, p)
    m = s
    t2 = 0
    while (t - 1) % p != 0:
        t2 = (t * t) % p
        for i in range(1, m):
            if (t2 - 1) % p == 0:
                break
            t2 = (t2 * t2) % p
        b = pow(c, 1 << (m - i - 1), p)
        r = (r * b) % p
        c = (b * b) % p
        t = (t * c) % p
        m = i
    return (True, r)


def isQuadraticResidue(p, n):
    # type: (prime, nonnegative) -> bool
    success, _modular_sqrt = tonelli(n, p)
    return success


def quadraticSieve(n, interval_mult=2, max_prime=229, verbosity=0):
    # type: (greater_than_one, nonnegative, greater_than_one, nonnegative) -> Dict[maybe_prime, greater_than_zero]
    assert n > 1, 'type violation, expected n > 1'
    assert interval_mult > 0, 'type violation, expected interval_mult > 0'
    # 1. choose smoothness bound B
    # TODO how do we pick B?
    B = max_prime
    # `primes` here is commonly referred to as a "factor_base" in other
    # implementations of the quadratic sieve
    primes = [p for p in slowPrimes(B) if isQuadraticResidue(p, n)]
    if verbosity > 0:
        print('Found %d primes less than or equal to %d, max is %d' % (len(primes), max_prime, primes[-1]))
    if verbosity > 2:
        print('Primes: %s' % (str(primes)))

    # 2. find numbers that are B smooth
    # TODO how do we pick the threshold for bSmoothSieve?
    ints, squares = bSmoothList(primes, n, interval_mult * len(primes))
    if verbosity > 0:
        print('Found %d b-smooth squares out of %d ints' % (len(squares), interval_mult * len(primes)))
    if verbosity > 2:
        print('Ints: %s' % (str(ints)))

    # 3. factor numbers and generate exponent vectors
    # 4. apply some linear algebra
    indices, products = findSquareProduct(primes, squares)
    if verbosity > 0:
        print('Found %d solutions to the mod-2 matrix' % (len(products)))

    for i, product in enumerate(products):
        used_ints = None  # type: Optional[List[nonnegative]]
        if verbosity > 1:
            used_ints = []
            print('Checking gcd of %d' % (product))
        # TODO move this computation to findSquareProduct
        a = 1
        for j, bit in enumerate(indices[i]):
            if bit == 0:
                continue
            a *= ints[j]
            if used_ints is not None:
                used_ints.append(ints[j])
        if used_ints is not None:
            print('Used ints: %s' % (str(used_ints)))
        product_root = intSqrt(product)
        # 5. now we have a ** 2 mod n == b ** 2 mod n
        # TODO should this be max(..) - min(..)?
        divisor = gcd(a - product_root, n)
        if divisor == 1 or divisor == n:
            continue
        # equivalent to `other_divisor = n // divisor`
        other_divisor = gcd(a + product_root, n)
        # 6. now we have (x - y) * (x + y) mod n == 0
        if verbosity > 1:
            x = (divisor + other_divisor) / 2
            y = (divisor - other_divisor) / 2
            print('%d = x ** 2 - y ** 2 = (x - y) * (x + y), x = %d, y = %d' % (n, x, y))
        return {divisor: 1, other_divisor: 1}

    # No solution found!
    return {}


def main():
    # type: () -> None
    parser = argparse.ArgumentParser(description='Factors a number using ' +
        'a quadratic sieve (kinda)')
    parser.add_argument('n', type=int, help='The number to factor')
    # TODO add help for mult (aka interval_mult)
    parser.add_argument('--mult', type=int, default=1)
    parser.add_argument('--max_prime', type=int, default=229, help='The ' +
        'max prime number to use for deriving B-smoothness. 229 is the ' +
        '50th prime, and 541 is the 100th.')
    parser.add_argument('-v', '--verbose', default=0, action='count')
    args = parser.parse_args()

    assert args.n > 3, 'expected n > 3'
    assert args.mult >= 1, 'expected mult >= 1'
    assert args.max_prime >= 2, 'expected max_prime >= 2'

    print(quadraticSieve(args.n, args.mult, args.max_prime, args.verbose))


if __name__ == '__main__':
    main()

File: primes/quadratic_sieve.c
// quadratic_sieve.c
// Trevor Pottinger
// Tue Dec 31 12:52:12 PST 2019

// This is mostly interesting to see how much faster c is than python. The
// slowest step is finding b-smooth numbers. So even if I don't finish
// implementing this completely, as long as we can output the list of
// b-smooth numbers, then maybe it can hand the remaining bits off to python. 

// Compiling:
// gcc quadratic_sieve.c -lm
// clang -lm quadratic_sieve.c

#include <stdio.h>

#include "quadratic_sieve.h"

// Error codes
#define INCORRECT_NUMBER_OF_ARGS 1
#define N_TOO_SMALL 2
#define MULT_TOO_SMALL 3
#define NUM_PRIMES_TOO_SMALL 4
#define FAILED_TO_FACTOR 5

void printUsage(char *command) {
  printf("printUsage() %s\n", command);
  printf("  n, must be greater than 1\n");
  printf("  mult, must be greater than 0\n");
  printf("  num_primes, must be greater than 1\n");
}

// https://stackoverflow.com/questions/3437404/min-and-max-in-c
i64 min(i64 a, i64 b) {
  if (a <= b) {
    return a;
  }
  return b;
}

i64 modPow(i64 base, i64 exp, i64 mod) {
  // TODO assert(mod > 1)
  i64 result = 1;
  // Does this line make sense?
  base = base % mod;
  while (exp > 0) {
    // This is logically equivalent to `exp % 2`
    if ((exp & 0x1) == 1) {
      result = (result * base) % mod;
    }
    exp = exp >> 1;
    base = (base * base) % mod;
  }
  return result;
}

bool slowFillPrimes(greater_than_one count, prime *arr) {
  // TODO implement count primes
  arr[0] = make_prime(2);
  arr[1] = make_prime(3);
  for (i64 i = 2; i < greater_than_one_val(count); ++i) {
    i64 prev_prime = prime_val(arr[i - 1]);
    // Keep iterating until we find the next prime
    for (i64 j = prev_prime + 2; true; j += 2) {
      bool is_prime = true;
      for (i64 k = 0; k < i; ++k) {
        if (j % prime_val(arr[k]) == 0) {
          is_prime = false;
          break;
        }
      }
      if (is_prime) {
        arr[i] = make_prime(j);
        break;
      }
    }
  }
  return true;
}

// Based on https://rosettacode.org/wiki/Integer_roots#C
// This currently rounds down, but I want it to round up.
i64 squareRoot(greater_than_one n) {
  i64 c = 1;
  i64 d = (1 + greater_than_one_val(n)) / 2;
  i64 e = (d + greater_than_one_val(n) / d) / 2;
  while (c != d && c != e) {
    c = d;
    d = e;
    e = (e + greater_than_one_val(n) / e) / 2;
  }
  if (d < e) {
    return d;
  }
  return e;
}

bool isBSmooth(positive num_primes, prime *primes, i64 n) {
  for (i64 i = 0; i < positive_val(num_primes); ++i) {
    while (n % prime_val(primes[i]) == 0) {
      n /= prime_val(primes[i]);
    }
    if (n == 1) {
      return true;
    }
  }
  // Idea: b-smoothish numbers
  // If n is a square number (i.e. `n == sqrt(n)**2`) then we could lie here
  // because it's large prime factors are all even. This means they would
  // contribute zero to the row reduction.
  return false;
}

bool bSmoothList(
  positive num_primes,
  prime *primes,
  greater_than_one n,
  i64 num_ints,
  i64 max_num_ints,
  i64 *num_ints_found,
  i64 *ints,
  greater_than_one *squares
) {
  *num_ints_found = 0;
  // TODO round the square root up, so we don't have to check if `squared`
  // is negative in the loop below
  i64 root = squareRoot(n);
  for (i64 i = root; i < root + num_ints && *num_ints_found < max_num_ints; ++i) {
    // If we had multiple polynomials, this is would need changing
    i64 squared = i * i - greater_than_one_val(n);
    if (squared < 0) {
      continue;
    }
    if (!isBSmooth(num_primes, primes, squared)) {
      continue;
    }
    ints[*num_ints_found] = i;
    squares[*num_ints_found] = make_greater_than_one(squared);
    ++*num_ints_found;
  }
  return true;
}

bool slowFactor(
  positive num_primes,
  prime *primes,
  greater_than_one n,
  i64 *num_factors,
  prime *factors,
  i64 *exponents
) {
  if (greater_than_one_val(n) == 2) {
    *num_factors = 1;
    factors[0] = make_prime(2);
    exponents[0] = 1;
    return true;
  }

  *num_factors = 0;
  for (i64 i = 0; i < positive_val(num_primes); ++i) {
    if (greater_than_one_val(n) % prime_val(primes[i]) != 0) {
      continue;
    }
    factors[*num_factors] = primes[i];
    i64 exponent = 0;
    do {
      ++exponent;
      n = make_greater_than_one(greater_than_one_val(n) / prime_val(primes[i]));
    } while (greater_than_one_val(n) % prime_val(primes[i]) == 0);
    exponents[*num_factors] = exponent;
    ++*num_factors;
    if (greater_than_one_val(n) == 1) {
      return true;
    }
  }

  return false;
}

void printInts(i64 num_ints, i64 *ints) {
  printf("[");
  for (i64 i = 0; i < num_ints; ++i) {
    printf("%lld", ints[i]);
    if (i < num_ints - 1) {
      printf(", ");
    }
  }
  printf("]");
}

void printFactors(i64 num_factors, i64 *factors, i64 *exponents) {
  printf("[");
  for (i64 j = 0; j < num_factors; ++j) {
    printf("(%lld, %lld)", factors[j], exponents[j]);
    if (j < num_factors - 1) {
      printf(", ");
    }
  }
  printf("]");
}

// TODO finish implementing findSquareProducts
bool findSquareProducts(
  positive num_primes,
  prime *primes,
  i64 num_ints,
  greater_than_one *ints
) {
  // Should we assert num_ints > num_primes? Most implementations search for
  // num_ints = num_primes + 1. Supposedly, each additional int more than prime
  // improves the likelyhood that a solution will be found.
  i64 num_factors[num_ints];
  prime factors[num_ints][MAX_NUM_DISTINCT_PRIME_FACTORS];
  i64 exponents[num_ints][MAX_NUM_DISTINCT_PRIME_FACTORS];

  for (i64 i = 0; i < num_ints; ++i) {
    bool factored = slowFactor(
      num_primes,
      primes,
      ints[i],
      &num_factors[i],
      factors[i],
      exponents[i]
    );
    // printf("Factored %lld: ", ints[i]);
    // printFactors(num_factors[i], factors[i], exponents[i]);
    // printf("\n");
    if (!factored) {
      // This should never happen when all the `ints` are b-smooth, where
      // `b = primes[num_primes - 1]` (i.e. the max prime). We fail here
      // because the code below doesn't handle un-factored ints.
      printf("WTF! Unable to factor: %lld\n", greater_than_one_val(ints[i]));
      return false;
    }
    // rows.append((num_factors, factors, exponents));
  }

  // exponents = vectorize(rows, default=0)
  // even_exponents = matrixMap(exponents, lambda x: x % 2)
  // TODO implement factors_to_col: prime -> column index
  i16 exponent_matrix[num_ints][positive_val(num_primes)];
  for (i64 i = 0; i < num_ints; ++i) {
    for (i64 j = 0; j < num_factors[i]; ++j) {
      i64 col_index = 0; // factors_to_col[factors[i][j]]
      exponent_matrix[i][col_index] = exponents[i][j] % 2;
    }
  }

  // TODO finish implementing findSquareProducts
  // result_matrix = rowReduce(matrixExtend(even_exponents, identity(count(rows))), 2)
  // foreach row in result_matrix:
  //   if isAllZeros(row[some_point:]):
  //     solutions.append(row)
  return false;
}

// Translated from https://rosettacode.org/wiki/Tonelli-Shanks_algorithm#Python
i64 legendre(i64 a, i64 p) {
  return modPow(a, (p - 1) / 2, p);
}

// Translated from https://rosettacode.org/wiki/Tonelli-Shanks_algorithm#Python
bool tonelli(i64 n, i64 p) {
  if (legendre(n, p) != 1) {
    return false;
  }
  // Really, this should return a tuple of (success, sqrt), but we only care
  // about if it succeeds. So we just return a bool.
  // TODO finish implementing the tonelli algorithm
  return true;
}

bool isQuadraticResidue(i64 p, i64 n) {
  // TODO filter primes to quadratic residues
  return tonelli(n, p);
}

bool quadraticSieve(
  greater_than_one n,
  positive mult,
  greater_than_one all_num_primes,
  i64 *results
) {
  prime all_primes[greater_than_one_val(all_num_primes)];
  bool success = slowFillPrimes(all_num_primes, all_primes);
  if (!success) {
    printf("Failed to find some small primes\n");
    return false;
  }

  prime primes[greater_than_one_val(all_num_primes)];
  i64 num_primes_found = 0;
  for (i64 i = 0; i < greater_than_one_val(all_num_primes); ++i) {
    if (!isQuadraticResidue(prime_val(all_primes[i]), greater_than_one_val(n))) {
      continue;
    }
    primes[num_primes_found] = all_primes[i];
    ++num_primes_found;
  }

  if (num_primes_found == 0) {
    printf("Failed to find find any primes\n");
    return false;
  }
  printf(
    "Successfully found %lld primes, max = %lld, out of all %lld primes considered\n",
    num_primes_found,
    prime_val(primes[num_primes_found - 1]),
    greater_than_one_val(all_num_primes)
  );
  positive num_primes = make_positive(num_primes_found);

  printf("Primes: ");
  printInts(num_primes_found, (i64*)(primes));
  printf("\n");

  // find b-smooth squares and their respective ints using n and primes
  i64 num_search_ints = positive_val(mult) * num_primes_found;
  i64 max_num_ints = min(num_search_ints, num_primes_found + 100);
  i64 ints[max_num_ints];
  greater_than_one squares[max_num_ints];
  i64 num_ints_found = -1;
  success = bSmoothList(
    num_primes,
    primes,
    n,
    num_search_ints,
    max_num_ints,
    &num_ints_found,
    ints,
    squares
  );

  if (!success) {
    printf("Failed to find a list of b-smooth squares\n");
    return false;
  }
  printf(
    "Found %lld b-smooth squares out of %lld ints, starting at %lld\n",
    num_ints_found,
    num_search_ints,
    ints[0]
  );

  if (num_ints_found == max_num_ints) {
    printf("Found the maximum number of b-smooth squares, rather than searching all ints\n");
  }
  printf("Ints: ");
  printInts(num_ints_found, ints);
  printf("\n");

  // find square products using primes and squares
  i64 num_solutions = -1;
  success = findSquareProducts(num_primes, primes, num_ints_found, squares);
  if (!success) {
    printf("Failed to find square products\n");
    return false;
  }

  // TODO check each solution

  return false;
}

int main(int argc, char** argv) {
  // Read the used command and decrement argc to be more readable
  char *command = argv[0];
  --argc;
  if (argc != 3) {
    printUsage(command);
    return INCORRECT_NUMBER_OF_ARGS;
  }

  // TODO parse an int larger than 64bits
  i64 n = atoll(argv[1]);
  if (n <= 1) {
    printUsage(command);
    return N_TOO_SMALL;
  }

  i64 mult = atoll(argv[2]);
  if (mult <= 0) {
    printUsage(command);
    return MULT_TOO_SMALL;
  }

  i64 num_primes = atoll(argv[3]);
  if (num_primes <= 1) {
    printUsage(command);
    return NUM_PRIMES_TOO_SMALL;
  }

  i64 factors[2];
  // All these type casts are based on the above bounds checking
  bool success = quadraticSieve(
    make_greater_than_one(n),
    make_positive(mult),
    make_greater_than_one(num_primes),
    factors
  );
  if (!success) {
    printf("Failed to factor %lld\n", n);
    return FAILED_TO_FACTOR;
  }

  return 0;
}
